### ISSUE TITLE
SQLAlchemy 2.0.0 takes forever to lint

### ISSUE THREAD

**Issue by chaen** [pylint-dev/pylint#8150]
### Bug description

Linting against `sqlalchemy` 2.0.0 is infinitely slower that against 1.4 series. 
This was reported here: https://github.com/sqlalchemy/sqlalchemy/issues/9181  

```python
# pylint: disable=missing-module-docstring

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine("")
session = sessionmaker(engine)
```

### Command used

```shell
time pylint slow.py
```

### Pylint output

```shell
With sqlalchemy 2.0.0


time pylint slow.py 

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 7.50/10, +2.50)


real    0m11.233s
user    0m11.080s
sys     0m0.112s
```

with 1.4.46

```
time pylint slow.py 

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)


real    0m4.149s
user    0m4.032s
sys     0m0.100s
```

On real life use cases, it goes from 6 seconds to 90 !
```


### Expected behavior

I would expect things to run faster

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.4
Python 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 15:55:03) 
[GCC 10.4.0]
```

### OS / Environment

Fedora 34, conda environment

### Additional dependencies

sqlalchemy

**Comment by clavedeluna** [comment]
Can easily reproduce on macOS Python 3.11.0. 

For some variety, I also ran it with `--enable=all --enable-all-extensions` and it's very similar results (so it's not 
like it's adding more checkers that's an issue). 

Running with `-verbose` doesn't add much info to this, other than it does hang longer at `Using config file ...`.

Next step, use a profiler.

**Comment by clavedeluna** [comment]
`python -m cProfile -m -o sql14 pylint test.py`
`python -m cProfile -m -o sql200 pylint test.py`

```
import pstats
from pstats import SortKey
fast = pstats.Stats('sql14')
slow = pstats.Stats('sql200')
fast.sort_stats(SortKey.CUMULATIVE).print_stats(10)
slow.sort_stats(SortKey.CUMULATIVE).print_stats(10)
```

```
>>> fast.sort_stats(SortKey.CUMULATIVE).print_stats(10)
Wed Feb  1 09:13:35 2023    sql14

         9883958 function calls (8618495 primitive calls) in 7.963 seconds

   Ordered by: cumulative time
   List reduced from 4010 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    423/1    0.010    0.000    7.971    7.971 {built-in method builtins.exec}
        1    0.000    0.000    7.971    7.971 <frozen runpy>:201(run_module)
        1    0.000    0.000    7.965    7.965 <frozen runpy>:65(_run_code)
        1    0.000    0.000    7.965    7.965 /Users/dani/pylint/pylint/__main__.py:1(<module>)
        1    0.000    0.000    7.965    7.965 /Users/dani/pylint/pylint/__init__.py:28(run_pylint)
        1    0.000    0.000    7.072    7.072 /Users/dani/pylint/pylint/lint/run.py:119(__init__)
        1    0.000    0.000    6.660    6.660 /Users/dani/pylint/pylint/lint/pylinter.py:653(check)
        1    0.000    0.000    6.641    6.641 /Users/dani/pylint/pylint/lint/pylinter.py:753(_lint_files)
        1    0.000    0.000    6.641    6.641 /Users/dani/pylint/pylint/lint/pylinter.py:776(_lint_file)
        1    0.000    0.000    6.641    6.641 /Users/dani/pylint/pylint/lint/pylinter.py:1054(check_astroid_module)
```

```
>>> slow.sort_stats(SortKey.CUMULATIVE).print_stats(10)
Wed Feb  1 09:16:29 2023    sql200

         28488458 function calls (22226628 primitive calls) in 22.853 seconds

   Ordered by: cumulative time
   List reduced from 4054 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    423/1    0.012    0.000   22.864   22.864 {built-in method builtins.exec}
        1    0.000    0.000   22.864   22.864 <frozen runpy>:201(run_module)
        1    0.000    0.000   22.856   22.856 <frozen runpy>:65(_run_code)
        1    0.000    0.000   22.856   22.856 /Users/dani/pylint/pylint/__main__.py:1(<module>)
        1    0.000    0.000   22.856   22.856 /Users/dani/pylint/pylint/__init__.py:28(run_pylint)
        1    0.000    0.000   21.888   21.888 /Users/dani/pylint/pylint/lint/run.py:119(__init__)
        1    0.000    0.000   21.438   21.438 /Users/dani/pylint/pylint/lint/pylinter.py:653(check)
        1    0.000    0.000   21.418   21.418 /Users/dani/pylint/pylint/lint/pylinter.py:753(_lint_files)
        1    0.000    0.000   21.418   21.418 /Users/dani/pylint/pylint/lint/pylinter.py:776(_lint_file)
        1    0.000    0.000   21.418   21.418 /Users/dani/pylint/pylint/lint/pylinter.py:1054(check_astroid_module)
```

**Pylint is spending 3x more cumulative time analyzing slqalchemy 2.xx than 1.4.xx**

**Comment by Zahlii** [comment]
Is there a temporary workaround for this? Is there a way to tell astroid / pylint to ignore this module entirely? Or is the best bet to uninstall sqlalchemy during builds, run pylint, re-install, and ignore import errors?

**Comment by Pierre-Sassoulas** [comment]
The best we have right now is:
```
# Control the amount of potential inferred values when inferring a single object.
# This can help the performance when dealing with large functions or complex,
# nested conditions.
limit-inference-results = 100

# This flag controls whether pylint should warn about no-member and similar
# checks whenever an opaque object is returned when inferring. The inference
# can return multiple potential results while evaluating a Python object, but
# some branches might not be evaluated, which results in partial inference. In
# that case, it might be useful to still emit no-member and other checks for
# the rest of the inferred objects.
ignore-on-opaque-inference=yes

```
I'm not sure that it's going to help in this case because you might need to infer a single value but the recursion involved to do that is catastrophically costly. Might be worth a try though.

(Sorry for the semi-related joke but I'd like to introduce a new concept that I would call "design by ChatGPT": You know that your lib is not perfect and you want to make it better with intuitive configuration, so you ask chatGPT: What's the way to prevent catastrophic inference recursion in astroid using configuration ? Then ChatGPT tell you that:
```
from astroid import conf

conf.MAX_INFERENCE_STEPS = 1000
conf.TYPING_HEURISTICS = False
conf.INFERENCE_TIMEOUT = 10

By using these configuration options, you can prevent catastrophic inference recursion in Astroid and improve the performance and accuracy of your static analysis.
```. 
And bam, here's your design. Joke aside we talked about imposing a limit on inference by timeout or recursion limit, but it's not done right now.)

**Comment by Zahlii** [comment]
<img width="889" alt="image" src="https://user-images.githubusercontent.com/218582/217756888-94b5ea36-d661-4e81-be90-52354eb52bfd.png">

None of these changes seem to have any impact on runtime and/or counter behavior above.

**Comment by Pierre-Sassoulas** [comment]
Opened #8248 to discuss a long term solution.

**Comment by Zahlii** [comment]
As a short-term workaround, I was able to pin sqlalchemy<2.0.0 via a constraints file and making sure to set the future=True argument (see also https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-to-2-0-step-four-use-the-future-flag-on-engine )

**Comment by tl24** [comment]
Does https://github.com/rucio/rucio/pull/6123 fix this issue? If so, is it possible to get an early build to use or test with until then? For us, one our pylint runs is taking about 30 minutes.

**Comment by kuba-lilz** [comment]
Just chipping in that for my fairy large project updating sqlalchemy from 1.4 to 2.0 bumps up pylint time from 30 seconds to over 6 minutes...

**Comment by mbyrnepr2** [comment]
I am seeing the Astroid module [brain_namedtuple_enum.py](https://github.com/pylint-dev/astroid/blob/main/astroid/brain/brain_namedtuple_enum.py) has a big affect on the results:

Standard Pylint run:
```python
time pylint example.py 
pylint example.py  20.82s user 0.68s system 95% cpu 22.510 total
```

With brain_namedtuple_enum.py removed from a local install of astroid:
```python
time pylint example.py 
pylint example.py  8.83s user 0.48s system 93% cpu 9.977 total
```

```python
SQLAlchemy==2.0.0
Darwin Kernel Version 20.6.0
```

**Comment by kuba-lilz** [comment]
Sure happy to hear that root cause was fixed, but until it's released I don't think it's correct to close this issue. I can install latest pylint and astroid and sqlalchemy linting is still slow.

**Comment by jacobtylerwalls** [comment]
We close issues when merging fixes, not when cutting releases.

**Comment by jacobtylerwalls** [comment]
You can choose to install bleeding-edge astroid or wait for an alpha or a stable version, it's up to each user.

**Comment by jacobtylerwalls** [comment]
The expected release milestone is marked in this issue.

**Comment by POD666** [comment]
I have been waiting for releases for a while, seems to be the latest
```
astroid==2.15.6
pylint==2.17.5
sqlalchemy==2.0.5
```
but linting time is still extremely huge
```
> time pylint [single-file]
...
pylint   33.78s
```
Do I miss anything? Which version of astroid/pylint is expected to contain the fix?

**Comment by POD666** [comment]
Ah, looks like `astroid==2.15.6` doesn't contain your fix. Right?

Is it expected in >3 and not sooner?

**Comment by POD666** [comment]
Looks like `pylint==3.0.0a6` and `astroid==3.0.0a8` doesn't work together 🥲
```
File "/Users/viktordanyliuk/.pyenv/versions/3.10.7/envs/venv-databand-py310/lib/python3.10/site-packages/pylint/checkers/imports.py", line 1038, in ImportsChecker
    @astroid.decorators.cached
AttributeError: module 'astroid.decorators' has no attribute 'cached'
```

**Comment by jacobtylerwalls** [comment]
Yep, support for the latest astroid improvements only exists on the main branch until we can get a release out. You can install from the main branch with `pip`, though.

**Comment by jacobtylerwalls** [comment]
You'll notice there's only one issue left on the 3.0.0a7 milestone, which another maintainer regards as a release blocker.
